File: fixtures/app/services/auth_service.py
"""
Service d'authentification.
Démontre:
- Relation circulaire avec user_service
- Dépendances stdlib (hashlib)
- Relations internes vers utils
"""

import hashlib
import secrets
from typing import Optional

# Relation INTERNE - vers utils
from ..utils.helpers import log_action

# RELATION CIRCULAIRE avec user_service
# Note: Import différé pour éviter l'erreur d'import circulaire
# Dans une vraie analyse, cette relation doit être détectée


# Stockage en mémoire des tokens (simulation)
_active_tokens: dict[str, int] = {}


def hash_password(password: str) -> str:
    """
    Hash un mot de passe de manière sécurisée.
    
    Relations:
    - Utilise hashlib (stdlib)
    """
    salt = secrets.token_hex(16)
    hashed = hashlib.pbkdf2_hmac(
        'sha256',
        password.encode('utf-8'),
        salt.encode('utf-8'),
        100000
    )
    return f"{salt}${hashed.hex()}"


def verify_password(password: str, stored_hash: str) -> bool:
    """
    Vérifie un mot de passe contre son hash stocké.
    
    Relations:
    - Utilise hashlib (stdlib)
    """
    try:
        salt, hash_value = stored_hash.split('$')
        new_hash = hashlib.pbkdf2_hmac(
            'sha256',
            password.encode('utf-8'),
            salt.encode('utf-8'),
            100000
        )
        return new_hash.hex() == hash_value
    except ValueError:
        return False


def authenticate(email: str, password: str) -> Optional[str]:
    """
    Authentifie un utilisateur et retourne un token.
    
    Relations:
    - Import différé de user_service (circulaire)
    - Utilise log_action (helpers)
    - Utilise secrets (stdlib)
    """
    # Import différé pour gérer la dépendance circulaire
    from .user_service import get_user
    
    log_action(f"Authentication attempt for {email}")
    
    # Simulation - en vrai on chercherait par email
    user = get_user(user_id=1)
    
    if user and user.email == email:
        # Générer un token
        token = secrets.token_urlsafe(32)
        _active_tokens[token] = user.id
        log_action(f"User {email} authenticated successfully")
        return token
    
    log_action(f"Authentication failed for {email}")
    return None


def verify_token(token: str) -> bool:
    """
    Vérifie si un token est valide.
    
    Relations:
    - Utilise log_action (helpers)
    """
    is_valid = token in _active_tokens
    log_action(f"Token verification: {'valid' if is_valid else 'invalid'}")
    return is_valid


def logout(token: str) -> bool:
    """
    Invalide un token.
    
    Relations:
    - Utilise log_action (helpers)
    """
    if token in _active_tokens:
        del _active_tokens[token]
        log_action("User logged out")
        return True
    return False

File: fixtures/php/src/Services/AuthService.php
<?php
/**
 * Service d'authentification
 */

declare(strict_types=1);

namespace App\Services;

require_once __DIR__ . '/../Models/User.php';
require_once __DIR__ . '/../Utils/helpers.php';

use App\Models\User;
use DateTime;
use function App\Utils\generateId;

/**
 * Gère l'authentification des utilisateurs
 */
class AuthService
{
    /** @var array<string, User> */
    private array $users = [];

    /** @var array<string, string> Token -> User ID */
    private array $sessions = [];

    public function __construct()
    {
        // Créer un utilisateur de test
        $testUser = new User(
            id: 'user-001',
            email: 'admin@example.com',
            name: 'Admin User'
        );
        $testUser->setPassword('admin123');
        $this->users[$testUser->getId()] = $testUser;
    }

    /**
     * Authentifie un utilisateur avec email et mot de passe
     */
    public function login(string $email, string $password): ?string
    {
        $user = $this->findUserByEmail($email);

        if ($user === null) {
            return null;
        }

        if (!$user->verifyPassword($password)) {
            return null;
        }

        // Créer une session
        $token = $this->generateToken();
        $this->sessions[$token] = $user->getId();

        // Mettre à jour le dernier login
        $user->setLastLogin(new DateTime());

        return $token;
    }

    /**
     * Déconnecte un utilisateur
     */
    public function logout(string $token): bool
    {
        if (!isset($this->sessions[$token])) {
            return false;
        }

        unset($this->sessions[$token]);
        return true;
    }

    /**
     * Vérifie si un token est valide
     */
    public function validateToken(string $token): ?User
    {
        if (!isset($this->sessions[$token])) {
            return null;
        }

        $userId = $this->sessions[$token];
        return $this->users[$userId] ?? null;
    }

    /**
     * Enregistre un nouvel utilisateur
     */
    public function register(string $email, string $password, string $name): ?User
    {
        // Vérifier si l'email existe déjà
        if ($this->findUserByEmail($email) !== null) {
            return null;
        }

        $user = new User(
            id: generateId(),
            email: $email,
            name: $name
        );
        $user->setPassword($password);

        $this->users[$user->getId()] = $user;
        return $user;
    }

    private function findUserByEmail(string $email): ?User
    {
        foreach ($this->users as $user) {
            if ($user->getEmail() === $email) {
                return $user;
            }
        }
        return null;
    }

    private function generateToken(): string
    {
        return bin2hex(random_bytes(32));
    }
}
